### DISCUSSION TITLE
Loaders and Vue SFCs

### DISCUSSION THREAD

**Discussion by cjpearson** [swc-project/swc#1519] [Q&A] [answered]
Hey there, I was playing around a bit with spack. As far as I understand, right now you cannot configure it to understand different filetypes. I'm interested in getting .vue files working. 

How do you think that should work? Should custom loaders written in JS be supported similar to webpack or will support have to be added directly to spack?

**Comment by kdy1** [comment, marked as answer]
Both are acceptable, but I think implementing it in rust would be better because it will show errors when required type definitions (ast) are changed.
Just like jest support, it can be implemented via some internal apis with `@swc/vue`.

**Reply by kdy1** [reply]
Regardless of how vue support is implemented, I think javascript-based custom loader is a must.

**Reply by cjpearson** [reply]
Yeah, that makes sense to me. Even if the vue parsing was handled within spack there would still need to be support for additional loaders if you use a different language. e.g. a `<style>` block that uses scss or less. 

I see that there's an issue for webpack-compatible plugins. Do you also have a plan for loaders? There's quite a lot you can do with webpack loaders, but maybe it makes sense to start with a simple API?

**Reply by kdy1** [reply]
Webpack-compatible plugins I meant include loader.
I also think loader is the first step, and I absolutely have plan to support them.

**Reply by cjpearson** [reply]
That's awesome. Is there any way I can help out?

**Reply by kdy1** [reply]
Maybe one of

 - applying javascript loaders (from rust)
 - creating webpack-compatible js object (from rust)
   - `this` object (loader context, https://webpack.js.org/api/loaders/#the-loader-context)
   - compiler object (https://webpack.js.org/api/loaders/#this_compiler)
   - compilation object (https://webpack.js.org/api/loaders/#this_compilation)

This is all the work that comes to mind now.
Maybe there is something more to do.

**Reply by cjpearson** [reply]
I can certainly take a crack at it. :) 

Are there any other places in swc where javascript is invoked from rust code that I could use as a model? Or is something like this the right idea? https://github.com/dwoznicki/swc/blob/3d6a7c6772c9979e1bb35bb12cebb7ca519ca7ac/native/src/bundle.rs#L159-L174

**Reply by kdy1** [reply]
> Are there any other places in swc where javascript is invoked from rust code 

Currently there's no one, but the code you mentioned seems like a correct approach. Thank you in advance!

**Reply by cjpearson** [reply]
So I'm just starting to take a look at this and I'll admit I'm a bit lost. Could you share how you invoke your local build of spack for testing?

**Reply by kdy1** [reply]
Can you share your error messages?

It's an educated guess, but you may need to run `npm install` on the project directory and `/ecmascript/transforms`.

**Reply by cjpearson** [reply]
I haven't even gotten that far yet. :) I've been able to run spack on a test project following the instructions here: https://swc.rs/docs/usage-spack-cli and I can compile the swc project locally, but I'm not sure how to point to my locally compiled build.

**Reply by kdy1** [reply]
I guess you need to run `npm link` from the swc directory and then from your project directory.

**Reply by cjpearson** [reply]
Ah got it now! I was also missing the npm build step. Thanks!

**Comment by deleonio** [comment]
![image](https://user-images.githubusercontent.com/6279703/139516991-0168a602-968b-40f8-bdf0-a81b6c5a9046.png)

```js
const SWC_LOADER_TS = {
    test: /\.ts$/,
    use: [
        {
            loader: 'swc-loader',
            options: {
                jsc: {
                    parser: {
                        syntax: 'typescript',
                        decorators: true
                    },
                },
            },
        },
    ],
};
```
